### PR TITLE
Common subexpression elimination, including simplifications.

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

File: libevmcore/Instruction.h [hunk 1]
```diff
@@ -237,7 +237,7 @@ enum Tier
 struct InstructionInfo
 {
 	std::string name;	///< The name of the instruction.
-	int additional;		///< Additional items required in memory for this instructions (only for PUSH).
+	int additional;		//a/< Additional items required in memory for this instructions (only for PUSH).
```
Comment (LianaHus) [inline libevmcore/Instruction.h]: a

File: libevmcore/ExpressionClasses.cpp [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+#include <boost/noncopyable.hpp>
+#include <libevmcore/Assembly.h>
+#include <libevmcore/CommonSubexpressionEliminator.h>
+
+using namespace std;
+using namespace dev;
+using namespace dev::eth;
+
+
+bool ExpressionClasses::Expression::operator<(const ExpressionClasses::Expression& _other) const
```
Comment (LianaHus) [inline libevmcore/ExpressionClasses.cpp]: const&

File: libevmcore/ExpressionClasses.cpp [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+	ExpressionClasses::Id id = tryToSimplify(exp);
+	if (id < m_representatives.size())
+		return id;
+
+	exp.id = m_representatives.size();
+	m_representatives.push_back(exp);
+	return exp.id;
+}
+
+string ExpressionClasses::fullDAGToString(ExpressionClasses::Id _id)
```
Comment (LianaHus) [inline libevmcore/ExpressionClasses.cpp]: the function seems to be const

File: libevmcore/ExpressionClasses.cpp [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+		{{Instruction::MOD, {X, 0}}, [=]{ return u256(0); }},
+		{{Instruction::MOD, {0, X}}, [=]{ return u256(0); }},
+		{{Instruction::AND, {X, 0}}, [=]{ return u256(0); }},
+		{{Instruction::OR, {X, ~u256(0)}}, [=]{ return ~u256(0); }},
+		// operations involving an expression and itself
+		{{Instruction::AND, {X, X}}, [=]{ return X; }},
+		{{Instruction::OR, {X, X}}, [=]{ return X; }},
+		{{Instruction::SUB, {X, X}}, [=]{ return u256(0); }},
+		{{Instruction::EQ, {X, X}}, [=]{ return u256(1); }},
+		{{Instruction::EQ, {X, X}}, [=]{ return u256(1); }},
```
Comment (LefterisJP) [inline libevmcore/ExpressionClasses.cpp]: duplicate

**Pull request merged**